app/components/audio-player/audio-player.tsx

```diff
@@ -104,6 +107,34 @@ export function AudioPlayer({
     loadAndPlay(fileUri, track.webUri, sound, { shouldPlay: true }, onPlaybackStatusUpdate)
   }, [fileUri, track.webUri, sound, onPlaybackStatusUpdate])
 
+  const currentDownloadState = useRef(downloadState)
+
+  useEffect(() => {
+    if (downloadState.status !== currentDownloadState.current.status) {
+      __DEV__ &&
+        console.log(
+          `we switched from ${currentDownloadState.current.status} to ${downloadState.status}`,
+        )
+    }
+
+    // TODO: Move this to the response of `useDownload`?
+    // TODO: Make use of subscribe and a listener which sends old + new state?
+    // This would avoid the keeping the state in a ref.
+    if (
+      currentDownloadState.current.status === "DOWNLOADING" &&
+      downloadState.status === "DOWNLOADED"
+    ) {
+      switchSource(downloadState.uri)
+    } else if (
+      currentDownloadState.current.status === "DOWNLOADED" &&
+      downloadState.status === "NOT_DOWNLOADED"
+    ) {
+      switchSource(track.webUri)
+    }
+
+    currentDownloadState.current = downloadState
```

Comment:
Make `useDownload` return the correct URI? Working with `currentDownloadState` seems like black magic to me.

Reply:
Yeah, that makes more sense. I'm in general not too happy about leaking the state (it includes `DownloadResumable` instance), but some kind of knowledge about the state is probably necessary to the user of `useDownload`. Maybe `isDownloading`, `isPaused`, ... flags are the way to go. The URI would be `undefined` (or `null`) unless the state is `DOWNLOADED`, right?

Reply:
> If you use the is... flags, then for the progress bar, you'll need to have another field progress that should only have a valid value when isDownloading and isPaused is true.

We could ensure that at compile time by something like:

```ts
{
  isDownloading: true,
  isPaused: false
  progress: number,
  ...
} | 
{ 
  isDownloading: false,
  isPaused: false,
  progress: undefined,
  ...
}
```

which would mean `progress: number | undefined` in general and of type `number` in branches like `if (isDownloading || isPaused)`.

I also think leaking `downloadResumable` for now is fine.

> The URI is then always usable directly and always the best choice regarding the current download status (file when available and web URI otherwise). I don't have the complete picture in my mind though to know if that could also pose problems.

Sounds nice, I'll give it a shot.

Reply:
> We could ensure that at compile time by something like

Oh, I'm not fluent with typescript union types, "one-value"-types (what I mean is `undefined` as type) and "discriminating flags" (what I mean is `isDownloading` and `isPaused`). Wouldn't have come up with that!

> Sounds nice, I'll give it a shot.

Hooray!

Reply:
> Why do you use `undefined` instead of `null`?

I think a year ago I stopped using `null` to only have a single "null type" (+/- external libraries) and it became a habit. I didn't even check the rest of the codebase, sorry. Should I use `null` here?

> Maybe useAudioSource should also come with a loading boolean and an error (with code and message or only an error code)?

The loading boolean is currently encoded as `undefined`. However: `useAudioSource` also returns `undefined` if given `track: undefined`. Allowing `useAudioSource(undefined)` was my attempt to avoid [conditional hooks](https://reactjs.org/docs/hooks-rules.html#only-call-hooks-at-the-top-level), but the attempt is flawed: The early return in `useAudioSource` before `useDownload` is still an incorrect usage of React hooks.
In order to fix this I need to move some stuff around: The background music becomes its own component which we can then render conditionally. Are you fine with me going on a small refactor spree to make `useAudioSource` work with the background music?

As for errors: Adding an error state to it sounds reasonable. What do you think about falling back to the web URI and only log the error (e.g. when accessing the file cache fails)?